packages/gatsby/src/utils/jobs-manager.js

```diff
@@ -0,0 +1,265 @@
+const uuid = require(`uuid/v4`)
+const path = require(`path`)
+const hasha = require(`hasha`)
+const fs = require(`fs-extra`)
+const pDefer = require(`p-defer`)
+const slash = require(`slash`)
+const _ = require(`lodash`)
+const { createContentDigest } = require(`gatsby-core-utils`)
+const reporter = require(`gatsby-cli/lib/reporter`)
+
+let activityForJobs = null
+let activeJobs = 0
+
+/** @type {Map<string, {id: string, deferred: pDefer.DeferredPromise<any>}>} */
+const jobsInProcess = new Map()
+
+/**
+ * We want to use absolute paths to make sure they are on the filesystem
+ *
+ * @param {string} filePath
+ * @return {string}
+ */
+const convertPathsToAbsolute = filePath => {
+  if (!path.isAbsolute(filePath)) {
+    throw new Error(`${filePath} should be an absolute path.`)
+  }
+
+  return slash(filePath)
+}
+/**
+ * Get contenthash of a file
+ *
+ * @param {string} path
+ */
+const createFileHash = path => hasha.fromFileSync(path, { algorithm: `sha1` })
+
+/**
+ * @typedef BaseJobInterface
+ * @property {string} name
+ * @property {string} outputDir,
+ * @property {Record<string, any>} args
+
+ * @typedef JobInputInterface
+ * @property {string[]} inputPaths
+ * @property {{name: string, version: string, resolve: string}} plugin
+
+ * @typedef InternalJobInterface
+ * @property {string} id
+ * @property {string} contentDigest
+ * @property {{path: string, contentDigest: string}[]} inputPaths
+ * @property {{name: string, version: string, resolve: string, isLocal: boolean}} plugin
+ *
+ * I know this sucks but this is the only way to do it properly in jsdoc..
+ * @typedef {BaseJobInterface & JobInputInterface} JobInput
+ * @typedef {BaseJobInterface & InternalJobInterface} InternalJob
+ */
+
+/** @type {pDefer.DeferredPromise<void>|null} */
+let hasActiveJobs = null
+
+/**
+ * @deprecated
+ * TODO: Remove for Gatsby v3 (compatibility mode)
+ */
+exports.jobsInProcess = jobsInProcess
+
+/**
+ * Get the local worker function and execute it on the user's machine
+ *
+ * @template T
+ * @param {function({ inputPaths: InternalJob["inputPaths"], outputDir: InternalJob["outputDir"], args: InternalJob["args"]}): T} workerFn
+ * @param {InternalJob} job
+ * @return Promise<T>
```

Comment:
```suggestion
 * @return {Promise<T>}
```
this seems to trip mdx up because it considers it as a description of `returns` and so it tries to parse it and fails as it's not valid (documentation plugin creates markdown nodes for jsdocs):

```
 ERROR #11321  PLUGIN

"gatsby-plugin-mdx" threw an error while running the onCreateNode lifecycle:

unknown: Expected corresponding JSX closing tag for <T> (7:17)

   5 |       {...props}
   6 |       components={components}>
>  7 | <p>{`Promise`}<T></p>
     |                  ^
   8 |     </MDXLayout>
   9 |   )
  10 | };undefined: unknown: Expected corresponding JSX closing tag for <T> (7:17)
```